internal/controller/dnsrecord_healthchecks.go

```diff
@@ -2,180 +2,161 @@ package controller
-				Name:             fmt.Sprintf("%s-%s-%s", dnsRecord.Spec.RootHost, dnsEndpoint.DNSName, address),
-				Host:             &dnsRecord.Spec.RootHost,
-				Path:             config.Endpoint,
-				Port:             config.Port,
-				Protocol:         config.Protocol,
-				FailureThreshold: config.FailureThreshold,
-			}
+	desiredProbes := buildDesiredProbes(dnsRecord, common.GetLeafsTargets(common.MakeTreeFromDNSRecord(dnsRecord), ptr.To([]string{})), allowInsecureCerts)
 
-			result := healthCheckReconciler.Reconcile(ctx, spec, dnsEndpoint, probeStatus, address)
-			results = append(results, result)
+	for _, probe := range desiredProbes {
+		// if one of them fails - health checks for this record are invalid anyway, so no sense to continue
+		if err := controllerruntime.SetControllerReference(dnsRecord, probe, r.Scheme); err != nil {
+			return err
 		}
-	}
-
-	result := r.reconcileHealthCheckStatus(results, dnsRecord)
-	return result
-}
-
-func (r *DNSRecordReconciler) getProbeStatus(address string, dnsRecord *v1alpha1.DNSRecord) *v1alpha1.HealthCheckStatusProbe {
-	if dnsRecord.Status.HealthCheck == nil || dnsRecord.Status.HealthCheck.Probes == nil {
-		return nil
-	}
-	for _, probeStatus := range dnsRecord.Status.HealthCheck.Probes {
-		if probeStatus.IPAddress == address {
-			return &probeStatus
+		if err := r.ensureProbe(ctx, probe, logger); err != nil {
+			return err
 		}
 	}
-
+	logger.Info("Healthecks reconciled")
 	return nil
 }
 
-func (r *DNSRecordReconciler) reconcileHealthCheckStatus(results []provider.HealthCheckResult, dnsRecord *v1alpha1.DNSRecord) error {
-	var previousCondition *metav1.Condition
-	probesCondition := &metav1.Condition{
-		Reason: "AllProbesSynced",
-		Type:   "healthProbesSynced",
-	}
+// DeleteHealthChecks deletes all v1alpha1.DNSHealthCheckProbe that have ProbeOwnerLabel of passed in DNSRecord
+func (r *DNSRecordReconciler) DeleteHealthChecks(ctx context.Context, dnsRecord *v1alpha1.DNSRecord) error {
+	logger := log.FromContext(ctx).WithName("healthchecks")
+	logger.Info("Deleting healthchecks")
 
-	var allSynced = metav1.ConditionTrue
+	healthProbes := v1alpha1.DNSHealthCheckProbeList{}
 
-	if dnsRecord.Status.HealthCheck == nil {
-		dnsRecord.Status.HealthCheck = &v1alpha1.HealthCheckStatus{
-			Conditions: []metav1.Condition{},
-			Probes:     []v1alpha1.HealthCheckStatusProbe{},
-		}
+	if err := r.List(ctx, &healthProbes, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			ProbeOwnerLabel: BuildOwnerLabelValue(dnsRecord),
+		}),
+		Namespace: dnsRecord.Namespace,
+	}); err != nil {
+		return err
 	}
 
-	previousCondition = meta.FindStatusCondition(dnsRecord.Status.HealthCheck.Conditions, "HealthProbesSynced")
-	if previousCondition != nil {
-		probesCondition = previousCondition
+	for _, probe := range healthProbes.Items {
+		logger.V(1).Info(fmt.Sprintf("Deleting probe: %s", probe.Name))
+		if err := r.Delete(ctx, &probe); err != nil {
+			return err
```

Comment:
should we gather the errors and attempt a delete on each of them to avoid leavng around ones we could have deleted due to an issue with another probe?